### ISSUE TITLE
Skaffold pipes logs from out of date pods

### ISSUE THREAD

**Issue by WilliamDenniss** [GoogleContainerTools/skaffold#458]
### Expected behavior
Skaffold only pipes logs output of up to date pods.

### Actual behavior
Skaffold pipes logs of all pods in the deployment, including ones in the deployment that are out of date.

### Information

```
Skaffold version: v0.4.0
Operating system: 10.13.3
Content of skaffold.yaml:
apiVersion: skaffold/v1alpha2
kind: Config
build:
  artifacts:
  - imageName: gcr.io/wdenniss-gke-test17/nodedemo
deploy:
  kubectl:
    manifests:
      - k8s-*
profiles:
  - name: gcb
    build:
      googleCloudBuild:
        projectId: wdenniss-gke-test17
```

### Steps to reproduce the behavior
  1. Start with a working skaffold setup (I used a simple NodeJS app)
  2. Intentionally introduce a runtime issue
  3. Skaffold will pipe the logs indicating the issue
  4. Fix the bug
  5. Skaffold rebuilds and re-deploys
  6. Skaffold will now still the output of the old pods until they're terminated (the desired behavior is to only pipe the output of the up to date pods, as the others are redundant).
 

### Logs

```

[node-demo-84d8ddf48d-qzx5h node-demo] TypeError: res.headder is not a function
[node-demo-84d8ddf48d-qzx5h node-demo]     at /app/server.js:9:9
[node-demo-84d8ddf48d-qzx5h node-demo]     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
[node-demo-84d8ddf48d-qzx5h node-demo]     at next (/app/node_modules/express/lib/router/route.js:137:13)
[node-demo-84d8ddf48d-qzx5h node-demo]     at Route.dispatch (/app/node_modules/express/lib/router/route.js:112:3)
[node-demo-84d8ddf48d-qzx5h node-demo]     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
[node-demo-84d8ddf48d-qzx5h node-demo]     at /app/node_modules/express/lib/router/index.js:281:22
[node-demo-84d8ddf48d-qzx5h node-demo]     at Function.process_params (/app/node_modules/express/lib/router/index.js:335:12)
[node-demo-84d8ddf48d-qzx5h node-demo]     at next (/app/node_modules/express/lib/router/index.js:275:10)
[node-demo-84d8ddf48d-qzx5h node-demo]     at expressInit (/app/node_modules/express/lib/middleware/init.js:40:5)
[node-demo-84d8ddf48d-qzx5h node-demo]     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
Starting build...
DEBU[0014] Running docker build: context: ., dockerfile: Dockerfile 
DEBU[0014] Checking base image node:6-alpine for ONBUILD triggers. 
DEBU[0014] Found onbuild triggers [] in image node:6-alpine 
INFO[0014] Found dependencies for dockerfile [package.json .] 
DEBU[0014] Recursively adding .                         
INFO[0014] deps [.git/objects/pack/pack-5b8e44bfd05cb85a056443391c800793c591514a.pack k8s-deploy.yaml views/index/index.ejs .git/info/exclude .git/logs/refs/remotes/gitlab/master .git/packed-refs .git/index .git/hooks/commit-msg.sample .git/hooks/post-update.sample .git/hooks/prepare-commit-msg.sample .git/logs/refs/heads/master server.js .dockerignore .git/config Dockerfile k8s-service.yaml .git/hooks/pre-applypatch.sample .git/refs/heads/master .git/refs/remotes/origin/HEAD .git/hooks/pre-commit.sample skaffold.yaml package.json .git/description .git/hooks/applypatch-msg.sample .git/hooks/pre-receive.sample .git/hooks/update.sample .git/logs/HEAD .git/logs/refs/remotes/origin/HEAD .git/objects/pack/pack-5b8e44bfd05cb85a056443391c800793c591514a.idx .git/HEAD .git/hooks/pre-push.sample .git/hooks/pre-rebase.sample .git/refs/remotes/gitlab/master public/app.css] 
DEBU[0014] Skipping excluded path: .git/objects/pack/pack-5b8e44bfd05cb85a056443391c800793c591514a.pack 
DEBU[0014] Skipping excluded path: .git/info/exclude    
DEBU[0014] Skipping excluded path: .git/logs/refs/remotes/gitlab/master 
DEBU[0014] Skipping excluded path: .git/packed-refs     
DEBU[0014] Skipping excluded path: .git/index           
DEBU[0014] Skipping excluded path: .git/hooks/commit-msg.sample 
DEBU[0014] Skipping excluded path: .git/hooks/post-update.sample 
DEBU[0014] Skipping excluded path: .git/hooks/prepare-commit-msg.sample 
DEBU[0014] Skipping excluded path: .git/logs/refs/heads/master 
DEBU[0014] Skipping excluded path: .dockerignore        
DEBU[0014] Skipping excluded path: .git/config          
DEBU[0014] Skipping excluded path: .git/hooks/pre-applypatch.sample 
DEBU[0014] Skipping excluded path: .git/refs/heads/master 
DEBU[0014] Skipping excluded path: .git/refs/remotes/origin/HEAD 
DEBU[0014] Skipping excluded path: .git/hooks/pre-commit.sample 
DEBU[0014] Skipping excluded path: .git/description     
DEBU[0014] Skipping excluded path: .git/hooks/applypatch-msg.sample 
DEBU[0014] Skipping excluded path: .git/hooks/pre-receive.sample 
DEBU[0014] Skipping excluded path: .git/hooks/update.sample 
DEBU[0014] Skipping excluded path: .git/logs/HEAD       
DEBU[0014] Skipping excluded path: .git/logs/refs/remotes/origin/HEAD 
DEBU[0014] Skipping excluded path: .git/objects/pack/pack-5b8e44bfd05cb85a056443391c800793c591514a.idx 
DEBU[0014] Skipping excluded path: .git/HEAD            
DEBU[0014] Skipping excluded path: .git/hooks/pre-push.sample 
DEBU[0014] Skipping excluded path: .git/hooks/pre-rebase.sample 
DEBU[0014] Skipping excluded path: .git/refs/remotes/gitlab/master 
Sending build context to Docker daemon  10.75kB
Step 1/5 : FROM node:6-alpine
 ---> 198eca0871d8
Step 2/5 : ADD package.json /app/package.json
 ---> Using cache
 ---> 716776acb108
Step 3/5 : RUN cd /app && npm install
 ---> Using cache
 ---> 2bc5b95cefee
Step 4/5 : COPY . /app/
 ---> Using cache
 ---> 43e5d3cdbeb9
Step 5/5 : CMD ["node", "/app/server.js"]
 ---> Using cache
 ---> d4052147fd97
Successfully built d4052147fd97
Successfully tagged afa4d7e87f121b16bb134d5b3e2dbc01:latest
Digest: afa4d7e87f121b16bb134d5b3e2dbc01:latest
Successfully tagged gcr.io/wdenniss-gke-test17/nodedemo:d4052147fd97ab52aea179bcf6dee71bf20b709db6fa030010bb9451ce34b01f
The push refers to repository [gcr.io/wdenniss-gke-test17/nodedemo]
d251fcff40af: Preparing
15bee9f1a9dd: Preparing
990be7953b8d: Preparing
6e36e6c868b3: Preparing
4e5965ce48d4: Preparing
e53f74215d12: Preparing
e53f74215d12: Waiting
6e36e6c868b3: Layer already exists
4e5965ce48d4: Layer already exists
990be7953b8d: Layer already exists
15bee9f1a9dd: Layer already exists
d251fcff40af: Layer already exists
e53f74215d12: Layer already exists
d4052147fd97ab52aea179bcf6dee71bf20b709db6fa030010bb9451ce34b01f: digest: sha256:413fdc586c327cb685853638ddb654acc8d7b0c1986fb6c528ed60515b3d80f1 size: 1577
Build complete in 2.020020771s
Starting deploy...
DEBU[0016] Deploying path: k8s-deploy.yaml              
Deploying k8s-deploy.yaml...
DEBU[0016] Applying manifest: apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "4"
  generation: 1
  labels:
    run: node-demo
  name: node-demo
  selfLink: /apis/extensions/v1beta1/namespaces//deployments/node-demo
spec:
  replicas: 4
  selector:
    matchLabels:
      run: node-demo
  strategy:
    rollingUpdate:
      maxSurge: 10
      maxUnavailable: 3
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        run: node-demo
    spec:
      containers:
      - command:
        - node
        - app/server.js
        image: gcr.io/wdenniss-gke-test17/nodedemo:d4052147fd97ab52aea179bcf6dee71bf20b709db6fa030010bb9451ce34b01f
        imagePullPolicy: IfNotPresent
        name: node-demo
        ports:
        - containerPort: 80
          protocol: TCP
        readinessProbe:
          failureThreshold: 10
          httpGet:
            path: /
            port: 80
            scheme: HTTP
          initialDelaySeconds: 3
          periodSeconds: 1
          successThreshold: 1
          timeoutSeconds: 1
        resources:
          limits:
            cpu: 20m
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
status: {}
 
DEBU[0016] Running command: [kubectl --context gke_wdenniss-gke-test17_us-central1-a_cluster-1 apply -f -] 
DEBU[0016] Command output: stdout deployment "node-demo" configured
, stderr: , err: %!s(<nil>) 
DEBU[0016] Deploying path: k8s-service.yaml             
Deploying k8s-service.yaml...
WARN[0016] image [gcr.io/wdenniss-gke-test17/nodedemo] is not used by the deployment 
DEBU[0016] Applying manifest: apiVersion: v1
kind: Service
metadata:
  creationTimestamp: null
  labels:
    run: node-demo
  name: node-demo
  selfLink: /api/v1/namespaces//services/node-demo
spec:
  ports:
  - port: 80
    protocol: TCP
    targetPort: 80
  selector:
    run: node-demo
  sessionAffinity: None
  type: LoadBalancer
status:
  loadBalancer: {}
 
DEBU[0016] Running command: [kubectl --context gke_wdenniss-gke-test17_us-central1-a_cluster-1 apply -f -] 
INFO[0016] Waiting for node-demo-7f457c48d9-9fqj6 to be ready 
INFO[0016] Waiting for node-demo-7f457c48d9-9fqj6 to be ready 
INFO[0016] Waiting for node-demo-7f457c48d9-ddlxv to be ready 
INFO[0016] Waiting for node-demo-7f457c48d9-rxsx2 to be ready 
INFO[0016] Waiting for node-demo-84d8ddf48d-kvb67 to be ready 
INFO[0016] Waiting for node-demo-84d8ddf48d-49t2z to be ready 
INFO[0016] Waiting for node-demo-84d8ddf48d-qzx5h to be ready 
INFO[0016] Waiting for node-demo-84d8ddf48d-2b57c to be ready 
INFO[0016] Waiting for node-demo-7f457c48d9-ddlxv to be ready 
INFO[0016] Waiting for node-demo-7f457c48d9-rxsx2 to be ready 
INFO[0016] Waiting for node-demo-7f457c48d9-9fqj6 to be ready 
INFO[0016] Waiting for node-demo-7f457c48d9-ddlxv to be ready 
INFO[0016] Waiting for node-demo-7f457c48d9-rxsx2 to be ready 
DEBU[0016] Command output: stdout service "node-demo" configured
, stderr: , err: %!s(<nil>) 
Deploy complete in 854.567105ms
Watching for changes...
[node-demo-84d8ddf48d-kvb67 node-demo] TypeError: res.headder is not a function
[node-demo-84d8ddf48d-kvb67 node-demo]     at /app/server.js:9:9
[node-demo-84d8ddf48d-kvb67 node-demo]     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
[node-demo-84d8ddf48d-kvb67 node-demo]     at next (/app/node_modules/express/lib/router/route.js:137:13)
[node-demo-84d8ddf48d-kvb67 node-demo]     at Route.dispatch (/app/node_modules/express/lib/router/route.js:112:3)
[node-demo-84d8ddf48d-kvb67 node-demo]     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
[node-demo-84d8ddf48d-kvb67 node-demo]     at /app/node_modules/express/lib/router/index.js:281:22
[node-demo-84d8ddf48d-kvb67 node-demo]     at Function.process_params (/app/node_modules/express/lib/router/index.js:335:12)
[node-demo-84d8ddf48d-kvb67 node-demo]     at next (/app/node_modules/express/lib/router/index.js:275:10)
[node-demo-84d8ddf48d-kvb67 node-demo]     at expressInit (/app/node_modules/express/lib/middleware/init.js:40:5)
[node-demo-84d8ddf48d-kvb67 node-demo]     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)

```

**Comment by dgageot** [comment]
We should be able to ignore logs from Terminating pods

**Comment by booninite** [comment]
I think this is fixed in `0.6.0`, at least, I'm not seeing the error anymore.

**Comment by dgageot** [comment]
\o/ Let's close the issue then
